File: backend/src/main/java/com/modulo/state/OperationalSchemas.java
package com.modulo.state;

/** Structural contracts for operational documents; namespace binding applies to all callers. */
final class OperationalSchemas {
  private OperationalSchemas() {}

  static String definition(String namespace, String id) {
    return switch (id) {
      case "modulo.todo" ->
          namespace.equals("todo-lists")
              ? "{\"type\":\"object\",\"required\":[\"id\",\"title\",\"list\",\"priority\",\"done\"],\"properties\":{\"id\":{\"type\":\"string\",\"minLength\":1,\"maxLength\":120},\"title\":{\"type\":\"string\",\"maxLength\":10000},\"list\":{\"type\":\"string\",\"maxLength\":10000},\"priority\":{\"enum\":[\"LOW\",\"MEDIUM\",\"HIGH\",\"URGENT\"]},\"done\":{\"type\":\"boolean\"},\"noteId\":{\"type\":\"integer\",\"minimum\":1},\"dueDate\":{\"type\":\"string\",\"maxLength\":10}}}"
              : null;
      case "modulo.time-entry" ->
          namespace.equals("zeiterfassung")
              ? "{\"type\":\"object\",\"required\":[\"id\",\"date\",\"engagement\",\"description\",\"minutes\",\"rateEur\",\"billable\",\"billed\"],\"properties\":{\"id\":{\"type\":\"string\",\"minLength\":1,\"maxLength\":120},\"date\":{\"type\":\"string\",\"maxLength\":10000},\"engagement\":{\"type\":\"string\",\"maxLength\":10000},\"description\":{\"type\":\"string\",\"maxLength\":10000},\"minutes\":{\"type\":\"number\",\"minimum\":0,\"maximum\":100000000.0},\"rateEur\":{\"type\":\"number\",\"minimum\":0,\"maximum\":1000000000000.0},\"billable\":{\"type\":\"boolean\"},\"billed\":{\"type\":\"boolean\"}}}"
              : null;
      case "modulo.expense" ->
          namespace.equals("euer-datev")
              ? "{\"type\":\"object\",\"required\":[\"id\",\"date\",\"vendor\",\"description\",\"netEur\",\"vatRate\",\"category\"],\"properties\":{\"id\":{\"type\":\"string\",\"minLength\":1,\"maxLength\":120},\"date\":{\"type\":\"string\",\"maxLength\":10000},\"vendor\":{\"type\":\"string\",\"maxLength\":10000},\"description\":{\"type\":\"string\",\"maxLength\":10000},\"netEur\":{\"type\":\"number\",\"minimum\":0,\"maximum\":1000000000000.0},\"vatRate\":{\"type\":\"number\",\"enum\":[0,7,19]},\"category\":{\"type\":\"string\",\"maxLength\":10000}}}"
              : null;
      case "modulo.expense.categories" ->
          namespace.equals("euer-datev")
              ? "{\"type\":\"array\",\"maxItems\":10000,\"items\":{\"type\":\"string\",\"maxLength\":500}}"
              : null;
      case "modulo.expense.exported-periods" ->
          namespace.equals("euer-datev")
              ? "{\"type\":\"array\",\"maxItems\":10000,\"items\":{\"type\":\"string\",\"maxLength\":500}}"
              : null;
      case "modulo.pipeline.stages" ->
          namespace.equals("kanban")
              ? "{\"type\":\"array\",\"maxItems\":10000,\"items\":{\"type\":\"string\",\"maxLength\":500}}"
              : null;
      case "modulo.retention.classes" ->
          namespace.equals("gobd-vault")
              ? "{\"type\":\"array\",\"maxItems\":1000,\"items\":{\"type\":\"object\",\"required\":[\"id\",\"label\",\"years\"],\"properties\":{\"id\":{\"type\":\"string\",\"minLength\":1,\"maxLength\":120},\"label\":{\"type\":\"string\",\"maxLength\":10000},\"years\":{\"type\":\"integer\",\"minimum\":0,\"maximum\":1000}}}}"
              : null;
      case "modulo.invoice.seller" ->
          namespace.equals("rechnung")
              ? "{\"type\":\"object\",\"required\":[\"name\",\"address\"],\"properties\":{\"name\":{\"type\":\"string\",\"maxLength\":10000},\"address\":{\"type\":\"string\",\"maxLength\":10000},\"taxNumber\":{\"type\":\"string\",\"maxLength\":10000},\"vatId\":{\"type\":\"string\",\"maxLength\":10000},\"iban\":{\"type\":\"string\",\"maxLength\":10000},\"email\":{\"type\":\"string\",\"maxLength\":10000}}}"
              : null;
      default -> null;
    };
  }

  static void validate(String id, String key, com.fasterxml.jackson.databind.JsonNode value) {
    if (id.equals("modulo.todo") || id.equals("modulo.time-entry") || id.equals("modulo.expense")) {
      String recordId = value.path("id").asText();
      if (!recordId.matches("[A-Za-z0-9_-][A-Za-z0-9_.-]{0,119}")
          || !key.equals("record." + recordId)) fail();
      String dateField = id.equals("modulo.todo") ? "dueDate" : "date";
      if (value.has(dateField)) {
        try {
          String date = value.get(dateField).asText();
          if (!java.time.LocalDate.parse(date).toString().equals(date)) fail();
        } catch (java.time.DateTimeException invalid) {
          fail();
        }
      }
    }
    if (id.equals("modulo.retention.classes")) {
      var ids = new java.util.HashSet<String>();
      for (var item : value) if (!ids.add(item.path("id").asText())) fail();
    }
    if (id.equals("modulo.expense.categories")
        || id.equals("modulo.expense.exported-periods")
        || id.equals("modulo.pipeline.stages")) {
      var strings = new java.util.HashSet<String>();
      for (var item : value) {
        if (!strings.add(item.asText())) fail();
        if (id.equals("modulo.expense.exported-periods")) {
          try {
            java.time.YearMonth.parse(item.asText());
          } catch (java.time.DateTimeException invalid) {
            fail();
          }
        }
      }
    }
  }

  private static void fail() {
    throw new org.springframework.web.server.ResponseStatusException(
        org.springframework.http.HttpStatus.BAD_REQUEST, "STATE_SCHEMA_MISMATCH");
  }
}

File: frontend/src/features/workspace/operationalSchemas.ts
import type { TodoItem } from './todos';
import type { TimeEntry } from './timeTracking';
import type { ExpenseRecord } from './euer';
import type { RetentionClass } from './gobd';
import type { SellerProfile } from './invoicing';
import type { OperationalCollection } from './operationalState';

type ObjectValue = Record<string, unknown>;
const invalid = (): never => { throw new Error('Invalid operational record. Export the source for recovery.'); };
function object(value: unknown): ObjectValue {
  if (!value || typeof value !== 'object' || Array.isArray(value)) return invalid();
  return value as ObjectValue;
}
function text(value: unknown, max = 10000): string {
  if (typeof value !== 'string' || value.length > max) return invalid();
  return value;
}
function id(value: unknown): string {
  const result = text(value, 120);
  if (!/^[A-Za-z0-9_-][A-Za-z0-9_.-]*$/.test(result)) return invalid();
  return result;
}
function date(value: unknown): string {
  const result = text(value, 10);
  if (!/^\d{4}-\d{2}-\d{2}$/.test(result) || !Number.isFinite(Date.parse(result))
    || new Date(result).toISOString().slice(0, 10) !== result) return invalid();
  return result;
}
function number(value: unknown, maximum = 1e12): number {
  if (typeof value !== 'number' || !Number.isFinite(value) || value < 0 || value > maximum) return invalid();
  return value;
}
function boolean(value: unknown): boolean { if (typeof value !== 'boolean') return invalid(); return value; }
export function validateTodo(value: unknown): TodoItem {
  const item = object(value); id(item.id); text(item.title); text(item.list); boolean(item.done);
  if (!['LOW', 'MEDIUM', 'HIGH', 'URGENT'].includes(String(item.priority))) return invalid();
  if (item.dueDate !== undefined) date(item.dueDate);
  if (item.noteId !== undefined && (!Number.isSafeInteger(item.noteId) || Number(item.noteId) < 1)) return invalid();
  return { ...item } as unknown as TodoItem;
}
export function validateTimeEntry(value: unknown): TimeEntry {
  const item = object(value); id(item.id); date(item.date); text(item.engagement); text(item.description);
  number(item.minutes, 1e8); number(item.rateEur); boolean(item.billable); boolean(item.billed);
  return { ...item } as unknown as TimeEntry;
}
export function validateExpense(value: unknown): ExpenseRecord {
  const item = object(value); id(item.id); date(item.date); text(item.vendor); text(item.description);
  number(item.netEur); text(item.category);
  if (![0, 7, 19].includes(Number(item.vatRate)) || typeof item.vatRate !== 'number') return invalid();
  return { ...item } as unknown as ExpenseRecord;
}
export function validateRetentionClass(value: unknown): RetentionClass {
  const item = object(value); id(item.id); text(item.label);
  if (!Number.isSafeInteger(item.years)) return invalid(); number(item.years, 1000);
  return { ...item } as unknown as RetentionClass;
}
export function validateSeller(value: unknown): SellerProfile | null {
  if (value === null) return null;
  const item = object(value);
  for (const field of ['name', 'address']) text(item[field]);
  for (const field of ['taxNumber', 'vatId', 'iban', 'email']) if (item[field] !== undefined) text(item[field]);
  return { ...item } as unknown as SellerProfile;
}
export function validateStrings(value: unknown): string[] {
  if (!Array.isArray(value)) return invalid();
  const strings = value.map(item => text(item, 500));
  if (strings.length > 10000 || new Set(strings).size !== strings.length) return invalid();
  return strings;
}
export const TODO_COLLECTION: OperationalCollection<TodoItem> = {
  namespace: 'todo-lists', schemaId: 'modulo.todo', legacyKey: 'modulo-todos', validate: validateTodo,
};
export const TIME_COLLECTION: OperationalCollection<TimeEntry> = {
  namespace: 'zeiterfassung', schemaId: 'modulo.time-entry', legacyKey: 'modulo-time-entries', validate: validateTimeEntry,
};
export const EXPENSE_COLLECTION: OperationalCollection<ExpenseRecord> = {
  namespace: 'euer-datev', schemaId: 'modulo.expense', legacyKey: 'modulo-euer-expenses', validate: validateExpense,
};
